src/code/z_camera.c

```diff
@@ -1,3069 +1,6777 @@
-        phi_v0 = 1;
+    if (camera->playerGroundY == camera->playerPosRot.pos.y || camera->player->actor.gravity > -0.1f ||
+        camera->player->stateFlags1 & 0x200000) {
+        *yPosOffset = Camera_LERPCeilF(playerPosRot->pos.y, *yPosOffset, PCT(OREG(43)), 0.1f);
+        yPosDelta = playerPosRot->pos.y - *yPosOffset;
+        tmpPos0.y -= yPosDelta;
+        Camera_LERPCeilVec3f(&tmpPos0, &camera->posOffset, camera->yOffsetUpdateRate, camera->xzOffsetUpdateRate, 0.1f);
     } else {
-        phi_v0 = 0;
+        if (!(flags & FLG_OFFGROUND)) {
+            yPosDelta = playerPosRot->pos.y - *yPosOffset;
+            eyeAtDist = OLib_Vec3fDistXZ(at, &camera->eye);
+            phi_f16 = eyeAtDist;
+            Math_atan2f(yPosDelta, eyeAtDist);
+            temp_f0_2 = Math_tanf(DEG_TO_RAD(camera->fov * 0.4f)) * phi_f16;
+            if (temp_f0_2 < yPosDelta) {
+                *yPosOffset = *yPosOffset + (yPosDelta - temp_f0_2);
+                yPosDelta = temp_f0_2;
+            } else if (yPosDelta < -temp_f0_2) {
+                *yPosOffset = *yPosOffset + (yPosDelta + temp_f0_2);
+                yPosDelta = -temp_f0_2;
+            }
+            tmpPos0.y = tmpPos0.y - yPosDelta;
+        } else {
+            yPosDelta = playerPosRot->pos.y - *yPosOffset;
+            temp_f0_2 = Math_atan2f(yPosDelta, OLib_Vec3fDistXZ(at, &camera->eye));
+
+            if (temp_f0_2 > DEG_TO_RAD(OREG(32))) {
+                phi_f16 = 1.0f - sinf(temp_f0_2 - DEG_TO_RAD(OREG(32)));
+            } else if (temp_f0_2 < DEG_TO_RAD(OREG(33))) {
+                phi_f16 = 1.0f - sinf(DEG_TO_RAD(OREG(33)) - temp_f0_2);
+            } else {
+                phi_f16 = 1.0f;
+            }
+            tmpPos0.y -= (yPosDelta * phi_f16);
+        }
+
+        Camera_LERPCeilVec3f(&tmpPos0, &camera->posOffset, PCT(OREG(29)), PCT(OREG(30)), 0.1f);
+        camera->yOffsetUpdateRate = PCT(OREG(29));
+        camera->xzOffsetUpdateRate = PCT(OREG(30));
     }
 
-    if (phi_v0 != 0) {
-        sp30.pos.y -= 49.f;
-        *arg3 = func_8004389C(sp30.pos.y, *arg3, 0.1f, 0.2f);
-        camera->unk_100 = func_8004389C(0.4f, camera->unk_100, 0.2f, 0.02f);
+    tmpPos1.x = playerPosRot->pos.x + camera->posOffset.x;
+    tmpPos1.y = playerPosRot->pos.y + camera->posOffset.y;
+    tmpPos1.z = playerPosRot->pos.z + camera->posOffset.z;
+    Camera_LERPCeilVec3f(&tmpPos1, at, camera->atLERPStepScale, camera->atLERPStepScale, 0.2f);
+    return 1;
+}
+
+s32 Camera_CalcAtForHorse(Camera* camera, VecSph* eyeAtDir, f32 yOffset, f32* yPosOffset, s16 calcSlope) {
+    Vec3f* at = &camera->at;
+    Vec3f posOffsetTarget;
+    Vec3f atTarget;
+    s32 pad;
+    s32 pad2;
+    f32 playerHeight;
+    Player* player;
+    PosRot horsePosRot;
+
+    playerHeight = Player_GetHeight(camera->player);
+    player = camera->player;
+    func_8002EF44(&horsePosRot, player->rideActor);
+
+    if (((EnHorse*)player->rideActor)->unk_1F0 & 4 ? 1 : 0) {
```

Comment:
I suggest adding a macro for this check in `z_en_horse.h` like I did with EN_HORSE_CHECK_1 and such